### ISSUE TITLE
feature request: delay restart before status FATAL

### ISSUE THREAD

**Issue by guettli** [Supervisor/supervisor#487]
Use case:

during a software update the server can't restart since some needed python modules are not available yet. Supervisor retries N times (N comes from config. AFAIK defaults to 3).

After N fails the state FATAL gets entered.

In my use case the programm entered state FATAL although just some seconds later the restart would have been successfull.

Can you understand this use case?

A possible solution would be to use a delay in unit "seconds": After N failed retries wait M seconds and then try again. In my case this should be done to infinity (endless loop).

**Comment by aztlan2k** [comment]
:+1:   this would be really useful.

allowing us to configure how long between retries would be great.  Allowing for an backoff would also help.    

i think it would be great if we could also specify a retry count of infinity (never give up) in combination with a sane value for the `delay_betwen_retries`.

**Comment by mmh** [comment]
+1, would be useful.

**Comment by willybarro** [comment]
:+1: would be great. There's already a PR with that feature https://github.com/Supervisor/supervisor/pull/509, but it's not automergeable, and there's no discussion on it.

**Comment by mrook** [comment]
+1

**Comment by nicwest** [comment]
+1

**Comment by arinto** [comment]
+1

**Comment by nereusz** [comment]
+1 This would be very useful for me

**Comment by x0e-foobar** [comment]
+1

**Comment by MatthiasBach** [comment]
+1

**Comment by jsmirnov** [comment]
+1

**Comment by guettli** [comment]
Has someone enough knowledge to create patch for this?

**Comment by dbpolito** [comment]
I would love to see this too, so :+1:

**Comment by detailyang** [comment]
+1

**Comment by oryband** [comment]
:+1:

**Comment by vvanscherpenseel** [comment]
Would love to have this feature. :+1:

**Comment by fordnox** [comment]
+1

**Comment by siavashs** [comment]
👍

**Comment by guettli** [comment]
Why was this ticket closed? Was the feature request implemented? I can't see a code change on this issue or on #561.

**Comment by tonicospinelli** [comment]
+1

**Comment by oryband** [comment]
dear sir plz implement this alrdy k thx bai

**Comment by eggsby** [comment]
:thumbsup: would like this feature, thanks!

**Comment by 0x20h** [comment]
+1, would be useful for me too!

**Comment by caorong** [comment]
+1, hope this!

**Comment by conanfanli** [comment]
+1

**Comment by dieend** [comment]
+1

**Comment by miso-belica** [comment]
+1 for `delay_between_retries` and also exponential backoff would be great.

**Comment by pfuender** [comment]
+1

**Comment by toastbrotch** [comment]
+1

**Comment by pfuender** [comment]
Let's try with PR #659 is based on fclairamb's work and includes fixed for the tests.

**Comment by philpearl** [comment]
+1

**Comment by SilverBut** [comment]
It is quite useful! Can someone help with that?

**Comment by ikb42** [comment]
+1

**Comment by Doca** [comment]
+1

**Comment by kyle-long** [comment]
+1

**Comment by f4il** [comment]
+1

**Comment by draganHR** [comment]
+1

**Comment by ummae** [comment]
+1

**Comment by igama** [comment]
+1

delay_between_retries would be great

**Comment by dmarcantonio** [comment]
+1

**Comment by vlsd** [comment]
The [documentations](http://supervisord.org/subprocess.html#process-states) says:

> When an autorestarting process is in the `BACKOFF` state, it will be automatically restarted by supervisord. It will switch between `STARTING` and `BACKOFF` states until it becomes evident that it cannot be started because the number of `startretries` has exceeded the maximum, at which point it will transition to the `FATAL` state. Each start retry will take progressively more time.

Does that mean that this behavior is implemented but not made available? I am confused...

**Comment by igtw** [comment]
+1

**Comment by jonathan-kosgei** [comment]
Having this same issue using supervisor to run gogs on kubernetes, I need gogs to wait for mysql to init before crapping out
```
2016/08/18 21:20:33 [install.go:65 GlobalInit()] [E] Fail to initialize ORM engine: migrate: sync: dial tcp 127.0.0.1:3306: connection refused
2016-08-18 21:20:33,474 INFO exited: gogs (exit status 1; not expected)
2016-08-18 21:20:34,475 INFO gave up: gogs entered FATAL state, too many start retries too quickly
```

**Comment by mirfilip** [comment]
+1

**Comment by guicantoni** [comment]
+1

**Comment by ketchol** [comment]
+1

**Comment by cabloo** [comment]
Hacky implementation of this: sleep for x number of seconds after running your command. e.g.

```
[program:www]
command=bash -c "/var/www/bin/server.sh; sleep 3"
```

**Comment by Napas** [comment]
+1

**Comment by glagola** [comment]
+1

**Comment by johnmarcou** [comment]
+1

**Comment by gonesurfing** [comment]
+1

**Comment by sasounda** [comment]
+1

**Comment by rgrcnh** [comment]
+1

**Comment by csgyuricza** [comment]
👍

**Comment by csgyuricza** [comment]
The "sleep" alternative proposed by @cabloo does won't work because supervisord does not support multiple commands or quotes - not sure why.

**Comment by cabloo** [comment]
@csgyuricza looks like it's parsing the `;` wrong (cutting off at that point). I would try removing the space before it.

**Comment by csgyuricza** [comment]
@cabloo I tried without spaces as well; doesn't work. It's really annoying that supervisor doesn't allow you to concatenate 2 or more commands together

**Comment by cabloo** [comment]
@csgyuricza maybe try `||` instead of `;` (which will only sleep if the command fails)

**Comment by ssmulders** [comment]
@csgyuricza make sure you edit `directory=/bin/` as well, and then make the path in `-c "/absolute/path/to/script.sh; sleep 5s"` - absolute ;-)

Fired up just fine with me!

**Comment by jderusse** [comment]
I just use  `sh -c "/mycommand || (sleep 5s && false)"` to:
- sleep only when process failed
- exit with a >0 status code and let supervisor handle the error

**Comment by PotterRafed** [comment]
still waiting on this! +1

**Comment by davefarthing** [comment]
+1

**Comment by lucas-dall** [comment]
+1

**Comment by livesoftplock** [comment]
+3

**Comment by koprivajakub** [comment]
+1

**Comment by phaibin** [comment]
+1

**Comment by jimbrowne** [comment]
The BACKOFF state does apparently increase the delay between restart attempts.  The docs don't currently explain this.  c.f. https://github.com/Supervisor/supervisor/pull/659

Perhaps I'll find the time to submit a pull request for the docs.

**Comment by gomgomgom** [comment]
+1

**Comment by virusdefender** [comment]
+1

**Comment by bryantebeek** [comment]
+1

**Comment by yixiaol-m** [comment]
This is really needed for a process to auto recover from network issue or after external system has recovered.

**Comment by wardlawp** [comment]
+1

**Comment by garyelephant** [comment]
+1

**Comment by Melvin-mlp** [comment]
+1

**Comment by michallohnisky** [comment]
+1

**Comment by 0x20h** [comment]
The desired feature can already be achieved with:

```
[program:x]
...
autorestart=true
startretries=24 
```

This will already add an increasing delay for each retry.
I think this issue can be closed.

**Comment by diggerdu** [comment]
+1

**Comment by anatoliykim** [comment]
If you want fixed delay - make the program sleep before running.

```
[program:xxx]
command=sleep 30 && ...
```

**Comment by tchwpkgorg** [comment]
"command=sleep 30 && ..." doesn't work - it causes:

> sleep: invalid time interval ‘&&’
> sleep: invalid time interval ‘/usr/bin/php’
> sleep: invalid time interval ‘app/cron’
> sleep: invalid time interval ‘worker:runner’
> Try 'sleep --help' for more information.


They made it easy in systemd:

> Restart=always
> RestartSec=3

And it restarts, and waits 3 secs between restarts.

**Comment by tchwpkgorg** [comment]
While this one works:

> [program:www]
> command=bash -c "/var/www/bin/server.sh; sleep 3"

It causes quite a nasty output in process list (ps auxf), i.e.:

> www-data 13026  0.0  0.0   4512   852 ?        S    00:53   0:00  \_ sh -c sleep 10; /usr/bin/php worker.php
> www-data 13031  0.0  0.0  11736   764 ?        S    00:53   0:00  |   \_ sleep 10
> www-data 13027  0.0  0.0   4512   708 ?        S    00:53   0:00  \_ sh -c sleep 10; /usr/bin/php worker.php
> www-data 13033  0.0  0.0  11736   664 ?        S    00:53   0:00  |   \_ sleep 10
> www-data 13029  0.0  0.0   4512   756 ?        S    00:53   0:00  \_ sh -c sleep 10; /usr/bin/php worker.php
> www-data 13036  0.0  0.0  11736   820 ?        S    00:53   0:00  |   \_ sleep 10
> 

Also, it basically doubles the number of processes run by a given user. So if you have any monitoring checks which look at a number of processes run by a given user, you may need to make some bigger or smaller adjustments. Also, command line output will be different.

**Comment by cbj4074** [comment]
Several others have mentioned a "backoff" implementation and the `BACKOFF` state, but only [0x20h](https://github.com/Supervisor/supervisor/issues/487#issuecomment-328369726) mentioned this vital clue (though, I wish he'd been more explicit):

The mechanism that implements `startretries` does employ a backoff strategy that increases the delay by 1 second with each attempt.

In other words, you can set `startretries` to a value that is large enough to cover any "expected downtime" in your workflow without causing a self-inflicted DoS scenario.

While it might be nice to have control over the backoff computation, I concur that this has been addressed in as much as a self-inflicted DoS will not result from setting `startretries` too generously.

**Comment by nvictor** [comment]
can anyone points to where in the docs what 0x20h describes (the "add an increasing delay for each retry") is described? thanks.

**Comment by cbj4074** [comment]
@nvictor If it were documented, I doubt any of us would be here. ;) I discovered that the delay is increased by one second with each retry by examining `supervisord`'s log entries.

**Comment by rotorsolutions** [comment]
@nvictor @vlsd Has already pointed to this documentation:

http://supervisord.org/subprocess.html#process-states
> Each start retry will take progressively more time.

I was also looking for a solution for my problem. I connect to an IMAP server and get a connection timeout. Starting the script after a little delay works great, but supervisor is too quick to restart (for this job).

Therefor a delay as a configuration option would be great to prevent hacks like the 'sleep' option mentioned above (which I will try now as there is not configurable alternative)

**Comment by vlsd** [comment]
So it seems like *linear* backoff is implemented only (exponential backoff is the other common kind), and only at one, hard-coded rate (1s per retry, as per @cbj4074, but not actually documented). I no longer have a horse in the game.  It seems to me like providing the users with an ability to both switch between linear and exponential backoff and set the rate at which the backoff happens would be an ideal solution here. Failing that, documenting that the backoff is linear and set at a rate of 1 second per retry is also a solution. Simply mentioning that backoff happens, with no other details, is confusing.

**Comment by BibhuGlobussoft** [comment]
I am having the **FATAL state** issue due to which my supervisor goes down even if the supervisor is **running** but the **workers not processing any jobs**.
Did any body have any solution for this ?

**Comment by virusdefender** [comment]
Thanks to @jderusse

I write a new script which may help

```shell
#! /bin/bash
set -x

max_backoff=10
cmd=$1
backoff_file_name=/tmp/$(echo $cmd | md5sum).backoff

echo Running $cmd, backoff file name: $backoff_file_name

if [ ! -f $backoff_file_name ]
then
    echo -1 > $backoff_file_name
fi

seconds=$(cat $backoff_file_name)

if (($(($seconds+1)) > $max_backoff))
then
	seconds=$max_backoff
else
	seconds=$(($seconds+1))
fi

echo $seconds > $backoff_file_name

$cmd || (sleep $seconds && false)
```

 - If you are using Mac, md5 command should be `md5` instead of `md5sum`
 - Be sure to quote you command `./proc_wrapper.sh "ls -la"`
 - Adjust your own `$max_backoff`

**Comment by virusdefender** [comment]
Thanks @jderusse

```shell
#! /bin/bash
# set -x

max_backoff=5
cmd=$1
backoff_file_name=/tmp/$(echo $cmd | md5sum | cut -d " " -f1).backoff

echo Running $cmd, backoff file name: $backoff_file_name

$cmd

status=$?

if (( $status > 0))
then
        if [ ! -f $backoff_file_name ]
        then
            seconds=-1
        else
            seconds=$(cat $backoff_file_name)
        fi

        if (($(($seconds+1)) > $max_backoff))
        then
            seconds=$max_backoff
        else
            seconds=$(($seconds+1))
            echo $seconds > $backoff_file_name
        fi

    sleep $seconds

    echo Exit status $status, sleep $seconds
    exit 1
else
    rm $backoff_file_name
fi
```

I write a new shell to implement this function

 - use `md5` instead of `ms5sum` in MacOS
 - adjust your own `max_backoff`
 - quote the command `./tools/proc_wrapper.sh "ls /tmp/aaaa"`

**Comment by hiscal2015** [comment]
+1

**Comment by estshy** [comment]
+1

**Comment by bramstroker** [comment]
+1

**Comment by zhaodanwjk** [comment]
Has this problem been solved? How？

**Comment by Chupaka** [comment]
@zhaodanwjk current behaviour is the following: on first failure - restart after 1 second, on second failure - restart after 2 seconds, then 3 seconds, 4 seconds, etc. Until FATAL state.

**Comment by zhaodanwjk** [comment]
> @zhaodanwjk current behaviour is the following: on first failure - restart after 1 second, on second failure - restart after 2 seconds, then 3 seconds, 4 seconds, etc. Until FATAL state.

Thank you very much. We can only wait until the new function is developed or another method is adopted to solve it.

**Comment by Chupaka** [comment]
@zhaodanwjk  To solve what? What do you expect? Isn't `command=bash -c "runme.sh; sleep 3"` enough for you?

**Comment by zhaodanwjk** [comment]
@Chupaka Thank you very much, but I tried this method and could not solve my problem. I need to wait 100 seconds before restarting automatically when the service quits abnormally.

**Comment by jderusse** [comment]
@zhaodanwjk look at my previous comment https://github.com/Supervisor/supervisor/issues/487#issuecomment-292556138

**Comment by adrian-vlad** [comment]
A better workaround is to use a script that restarts fatal processes by configuring an eventlistener that receives **PROCESS_STATE_FATAL** events.

**Comment by icanhazstring** [comment]
I will link the PR for `symfony/messenger` here as they have addressed this using a wrapper script.
https://github.com/symfony/symfony-docs/pull/13597/files#diff-cefce3d23e1437ef5885ad7d486ec8d57efd7a3f6ad55bd5997bb202b537ff59R650
